glommio/src/sys/source.rs

```diff
@@ -225,3 +239,112 @@ impl Drop for Source {
         }
     }
 }
+
+pub struct SharedSource {
+    source: Source,
+    result: RefCell<Option<OsResult>>,
+    wakers: RefCell<VecDeque<Waker>>,
+}
+
+impl fmt::Debug for SharedSource {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("InnerSharedSource")
+            .field("source", &self.source)
+            .field("wakers", &self.wakers)
+            .field("result", &self.result)
+            .finish()
+    }
+}
+
+impl From<Source> for SharedSource {
+    fn from(src: Source) -> Self {
+        SharedSource {
+            source: src,
+            result: RefCell::new(None),
+            wakers: RefCell::new(Default::default()),
+        }
+    }
+}
+
+impl SharedSource {
+    pub(crate) fn source_type(&self) -> Ref<'_, SourceType> {
+        self.source.source_type()
+    }
+
+    pub(crate) fn buffer(&self) -> Ref<'_, IoBuffer> {
+        Ref::map(self.source_type(), |stype| match &*stype {
+            SourceType::Read(_, Some(buffer)) => buffer,
+            SourceType::Write(_, buffer) => buffer,
+            x => panic!("Could not extract buffer. Source: {:?}", x),
+        })
+    }
+
+    pub(crate) unsafe fn as_slice(&self) -> &[u8] {
+        match &*self.source_type() {
+            SourceType::Read(_, Some(buffer)) => {
+                std::slice::from_raw_parts(buffer.as_ptr(), buffer.len())
+            }
+            SourceType::Write(_, buffer) => {
+                std::slice::from_raw_parts(buffer.as_ptr(), buffer.len())
+            }
+            x => panic!("Could not extract buffer. Source: {:?}", x),
+        }
+    }
+
+    pub(crate) fn add_waiter(&self, waker: Waker) {
+        // add the waker to the underlying source and insert the previous one (if any)
+        // into the backlog.
+        if let Some(previous) = self.source.add_waiter(waker) {
+            // before inserting the waker make we replace an old one if any
```

Comment:
if any what?

I don't understand what are you trying to achieve with this code.

Reply:
`Source` can only have one waiter. Calling `add_waiter` replaces the current waiter and returns the previous one (or `None` if this is the first time it's called).

With `SharedSource` I want to create a queue of waiters. So when `add_waiter` is called, I first replace the waiter in the underlying `Source` and I add the previous one in a local vec so that I can't wake it up later on. This is what this code is doing.